File: Dictionaries and Sets/205. Isomorphic Strings/205. Isomorphic Strings.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  if (s.length !== t.length) return false;

  map1 = new Map();
  map2 = new Map();

  for (let i = 0; i < s.length; i++) {
    const char_s = s[i];
    const char_t = t[i];

    if (map1.has(char_s)) {
      if (map1.get(char_s) !== char_t) return false;
    } else {
      map1.set(char_s, char_t);
    }

    if (map2.has(char_t)) {
      if (map2.get(char_t) !== char_s) return false;
    } else {
      map2.set(char_t, char_s);
    }
  }

  return true;
};

// Time: O(n)
// Space: O(n)

File: Dictionaries and Sets/205. Isomorphic Strings/205. Isomorphic Strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        map1 = {}
        map2 = {}

        for char_s, char_t in zip(s, t):
            if char_s in map1:
                if map1[char_s] != char_t:
                    return False
            else:    
                map1[char_s] = char_t
            
            if char_t in map2:
                if map2[char_t] != char_s:
                    return False
            else:    
                map2[char_t] = char_s
        
        return True

# Time: O(n)
# Space: O(n)